doc/sfputil/dump_sfp_eeprom.md

```diff
@@ -0,0 +1,181 @@
+# Dump SFP EEPROM page data in show techsupport command #
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+SONiC show techsupport command provides the ability to collect system dump for debug purpose. Module EEPROM data is important information for PHY issue debugging, but it is not part of show techsupport command. This design will enhance show techsupport command to contain module EEPROM data.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+1. A new CLI `sfputil show eeprom-hexdump-all` shall dump EEPROM pages in a single call.
+2. Show techsupport command shall be extended to call `sfputil show eeprom-hexdump-all` command to collect module EEPROM data.
+
+### Requirements
+
+1. Show techsupport command shall collect module EEPROM raw data
+2. All existing cable types shall be supported except RJ45
+3. Vendor specific implementation is not required but still possible.
+
+### Architecture Design
+
+The current architecture is not changed
+
+### High-Level Design
+
+Submodule sonic-platform-common and sonic-utilities shall be changed. By default, vendor platform API implementation is not required.
+
+#### sonic-platform-common
+
+##### sfp_base change
+
+A new API `dump_eeprom` shall be added to `sfp_base.SfpBase`. This API shall be used by `sfputil show eeprom-hexdump-all` command to collect module EEPROM data.
+
+```python
+def dump_eeprom(self, page=None):
+    """
+    Dump all EEPROM data for this SFP
+
+    Args:
+        page: EEPROM page number, dump all pages if page is None
+
+    Returns:
+        A string contains the hex format EEPROM data
+    """
+    raise NotImplementedError
+```
+
+##### sfp_optoe_base change
+
+`sfp_optoe_base.SfpOptoeBase` shall implement the new API `dump_eeprom` so that vendor does not have to implement it (Vendor specific implementation is also possible).
+
+```python
+def dump_eeprom(self, page=None):
+    api = self.get_xcvr_api()
+    return api.dump_eeprom(page) if api is not None else None
+```
+
+##### sonic_xcvr change
+
+Following sonic_xcvr API implementation shall implement the new API `dump_eeprom`:
+
+- sonic_xcvr.api.public.cmis
+  - copper: page 0h (0-255)
+  - optical: pages 0h (0-255), 1h, 2h, 10h, 11h (128-255)
+- sonic_xcvr.api.public.sff8436
+  - copper: page 0h (0-255)
+  - optical: pages 0h (0-255), 1h, 2h, 3h (128-255)
+- sonic_xcvr.api.public.sff8472
+  - copper: page A0h (0-128)
+  - optical: page A0h (0-255), A2h (0-255)
+- sonic_xcvr.api.public.sff8636
+  - copper: page 0h (0-255)
+  - optical: pages 0h (0-255), 1h, 2h, 3h (128-255)
+
+#### sonic-utilities
+
+##### generate_dump change
+
+1. New subcommand `sfputil show eeprom-hexdump-all` shall be used to dump module EEPROM data
```

Comment:
@Junchao-Mellanox please add an option '--all' under eeprom-hexdump

Reply:
Hi @prgeor , I also preferred to add "--all", but I found that "sfputil show EEPROM-hexdump" requires a mandatory argument "--page", it seems that it is conflict with "--all". So, I created a new subcommand. Let me know if you think differently.

Reply:
@Junchao-Mellanox Lets make the page as optional argument and if user does not specify the page, then we display the pages from a list of hardcoded pages you are interested in

Reply:
> @Junchao-Mellanox Lets make the page as optional argument and if user does not specify the page, then we display the pages from a list of hardcoded pages you are interested in

@Junchao-Mellanox - Since you are planning to make page as optional argument, can we add `--all` option to dump EEPROM for all ports rather than directly using `sfputil show eeprom-hexdump`?

Reply:
Hi @mihirpat1, I had a offline discussion with Prince and we agreed with following:
```
1. `sfputil show eeprom-hexdump --port <port> --page <page>`: dump given page for given port
2. `sfputil show eeprom-hexdump --port <port>`: dump page 0 for given port (to keep backwardcompatible)
3. `sfputil show eeprom-hexdump --page <page>`: dump given page for all ports, validate that page must be in range [0,255]. User is repsonsible for making sure the page existence.
4. `sfputil show eeprom-hexdump`: dump available pages for all ports. Available pages for different cable types are described below.
```

Let me know if you think we need another short sync about this. I will schedule meeting if need.